lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -461,24 +386,35 @@ void TypeChecker::checkReferencedGenericParams(GenericContext *dc) {
   }
 }
 
-static GenericSignature *
-computeGenericFuncSignature(TypeChecker &tc, AbstractFunctionDecl *func) {
-  if (auto gp = func->getGenericParams()) {
+void TypeChecker::validateGenericFuncOrSubscriptSignature(
+                PointerUnion<AbstractFunctionDecl *, SubscriptDecl *>
```

Comment:
Another approach would be to pass in the ParameterList, TypeLoc for the return type and anything else it needs separately, together with the GenericContext that it will eventually call setGenericEnvironment() on (not saying it will necessarily be cleaner, but it might be worth a shot).

Reply:
We can actually derive anything from just the generic context: cast it to a subscript and a func, get the value declaration, the type locations and stuff, to save the callers from computing and passing a lot of parameters. I still absolutely need to differentiate between a function and a subscript in couple of places though. What I like about the current approach is type safety, i.e. you *have* to pass either a subscript or an abstract function, and only that. At the same time, the ugliness of the pointer union and condition statements is kind of a big `FIXME` comment that calls out to make functions and subscripts more related to each other. What do you think?
I believe accepting the «static subscripts» proposal (if we do) will be a decent impulse towards that.

Reply:
Right, I was just suggesting having the caller extract the fields so that you don’t have to cast here.

Reply:
Yes, I got you, what I meant is that we still need to cast to `SubscriptDecl` or `AbstractFuncDecl` in places where values can be considered implementation details the caller shouldn't know about:
```cpp
bool hasDynamicSelf = false;
// can only be true for a FuncDecl
```
```cpp
typeCheckParameterList(params, resolution, func
                                           ~~~~
                             ? TypeResolverContext::AbstractFunctionDecl
                             : TypeResolverContext::SubscriptDecl);
```
```cpp
func ? func->computeType() : subscr->computeType();
~~~~                         ~~~~~
```
Of course, I can pass those as parameters as well, including the `computeType` function, but I think it's an overkill, considering we would have around 8 parameters. 
*But* if you think differently let me know, I will give it a try.